gtp_cmd_handlers.py

```diff
@@ -0,0 +1,340 @@
+                                    was_resign=False)
+        return coords.to_kgs(move)
+
+    def cmd_undo(self):
+        raise NotImplementedError()
+
+    def cmd_final_score(self):
+        return self._player.get_result_string()
+
+    def _accomodate_out_of_turn(self, color: str):
+        position = self._player.get_position()
+        if translate_gtp_color(color) != position.to_play:
+            position.flip_playerturn(mutate=True)
+
+
+class KgsCmdHandler(object):
+    def __init__(self, player):
+        self._player = player
+
+    def cmd_time_left(self):
+        pass
+
+    def cmd_showboard(self):
+        print('\n\n' + str(self._player.get_position()) + '\n\n', file=sys.stderr)
+        return True
+
+    def cmd_kgs_chat(self, msg_type: str, sender: str, text: str):
+        if not hasattr(self._player, 'get_root'):
+            return "I have nothing interesting to say."
+
+        root = self._player.get_root()
+        default_response = "Supported commands are 'winrate', 'nextplay', 'fortune', and 'help'."
+        if root is None or root.position.n == 0:
+            return "I'm not playing right now.  " + default_response
+
+        if 'winrate' in text.lower():
+            wr = (abs(root.Q) + 1.0) / 2.0
+            color = "Black" if root.Q > 0 else "White"
+            return "{:s} {:.2f}%".format(color, wr * 100.0)
+        elif 'nextplay' in text.lower():
+            return "I'm thinking... " + root.most_visited_path()
+        elif 'fortune' in text.lower():
+            return "You're feeling lucky!"
+        elif 'help' in text.lower():
+            return "I can't help much with go -- try ladders!  Otherwise: " + default_response
+        else:
+            return default_response
+
+
+class RegressionsCmdHandler(object):
+    def __init__(self, player):
+        self._player = player
+
+    def cmd_loadsgf(self, filename: str, movenum=0):
+        try:
+            with open(filename, 'r') as f:
+                contents = f.read()
+        except:
+            raise ValueError("Unreadable file: " + filename)
+
+        try:
+            # This is kinda bad, because replay_sgf is already calling
+            # 'play move' on its internal position objects, but we really
+            # want to advance the engine along with us rather than try to
+            # push in some finished Position object.
+            for idx, p in enumerate(sgf_wrapper.replay_sgf(contents)):
+                print("playing #", idx, p.next_move, file=sys.stderr)
+                self._player.play_move(p.next_move)
+                if movenum and idx == movenum:
+                    break
+        except:
+            raise
+
+
+class GoGuiCmdHandler(object):
+    """GTP extensions of 'analysis commands' for gogui."""
+
+    def __init__(self, player):
+        self._player = player
+
+    def cmd_gogui_analyze_commands(self):
+        return "\n".join(["var/Most Read Variation/nextplay",
```

Comment:
are these parsed / interpreted somewhere?

Reply:
yeah, go gui adds the output of this command to its UI, giving you buttons or whatever